contracts/Removal.sol

```diff
@@ -80,57 +110,138 @@ contract Removal is
     bytes memory data
   ) public override {
+    uint256 idsLength = ids.length;
+    if (!(amounts.length == idsLength)) {
+      revert ArrayLengthMismatch({array1Name: "amounts", array2Name: "ids"});
+    }
+    // todo should we check that the removal id-encoded supplier address for each id is the same as `to` ?
+    // todo should we validate the removal ids any further? what about other malformed fields?
+    //    note that there is validation performed when creating ids in the first place, but room for errors/changes between id creation and then submission to this function
+    // todo do we add any validation to enforce that all removals in batch belong to the same project id?
     BatchMintRemovalsData memory decodedData = abi.decode(
       data,
       (BatchMintRemovalsData)
     );
-    for (uint256 i = 0; i < ids.length; i++) {
-      if (_tokenIdExists[ids[i]]) {
-        revert TokenIdExists(); // todo can the duplicate token id be reported here?
+    uint256 projectId = decodedData.projectId;
+    uint256 scheduleStartTime = decodedData.scheduleStartTime;
+    bool list = decodedData.list;
+    address marketAddress = decodedData.marketAddress;
+    if (projectId == 0) {
+      revert InvalidProjectId({projectId: projectId});
+    }
+    if (
+      scheduleStartTime < UNIX_EPOCH_2010 || scheduleStartTime > UNIX_EPOCH_2050
+    ) {
+      revert InvalidScheduleStartTime({startTime: scheduleStartTime});
+    }
+    uint256 newTokenIdCounter = tokenIdCounter;
+    for (uint256 i = 0; i < idsLength; i++) {
+      uint256 id = ids[i];
+      if (_tokenIdExists[id]) {
+        revert TokenIdExists({tokenId: id});
       }
-      _tokenIdExists[ids[i]] = true;
-      indexToTokenId[_tokenIdCounter] = ids[i];
-      _tokenIdCounter += 1;
+      _removalIdToProjectId[id] = projectId;
+      _tokenIdExists[id] = true;
+      indexToTokenId[newTokenIdCounter] = id;
+      newTokenIdCounter += 1;
     }
+    uint256 firstId = ids[0];
+    tokenIdCounter = newTokenIdCounter;
+    _projectIdToScheduleData[projectId] = ScheduleData(
+      scheduleStartTime,
+      firstId.supplierAddress(), // todo check gas costs of extracting values per line vs once using unpackRemovalIdV0
+      firstId.methodology(), // todo check gas costs of extracting values per line vs once using unpackRemovalIdV0
+      firstId.methodologyVersion() // todo check gas costs of extracting values per line vs once using unpackRemovalIdV0
+    );
     super.mintBatch(to, ids, amounts, data);
-
     setApprovalForAll(to, _msgSender(), true); // todo look at vesting contract for potentially better approach
-    if (decodedData.list) {
-      safeBatchTransferFrom(to, decodedData.marketAddress, ids, amounts, data);
+    if (list) {
+      safeBatchTransferFrom(to, marketAddress, ids, amounts, data);
     }
   }
 
+  function mint(
+    address,
+    uint256,
+    uint256,
+    bytes calldata
+  ) public pure override {
+    revert("Removal: ERC 1155 mint disabled");
```

Comment:
I don't like how this is exposed in the interface at all, and wonder if we should stop extending from the mintable preset (minting is an optional extension for 1155s-- not even listed in [the EIP interface spec](https://eips.ethereum.org/EIPS/eip-1155)-- so I dont _think_ it would hurt to just omit it all together so that it's not exposed on things like etherscan, etc)

Reply:
Where can i find that version?

Reply:
that's what we're currently looking at

Reply:
we don't inherit burning anymore because all we did was then disable it

Reply:
mint still exists though?

Reply:
no mint is also no longer inherited - stopped inheriting the `PresetMinterPauserUpgradeable` and just inherit `PausableUpgradeable` now